### ISSUE TITLE
"Pointer to a local variable escapes from the function" is wrong

### ISSUE THREAD

**Issue by skejeton** [vtereshkov/umka-lang#199]
```
type ToolNode = struct {
  children: map[str]^ToolNode
  parent: ^ToolNode
}

type Editor = struct {
  tools: ToolNode
  currentTool: ^ToolNode
}

fn (e: ^Editor) beginTool(name: str) {
  tool := new(ToolNode)
  tool ^= ToolNode{children: map[str]^ToolNode{}, parent: e.currentTool}
  e.currentTool.children[name] = tool
  e.currentTool = e.currentTool.children[name]
}

fn main() {
  e := Editor{tools: ToolNode{children: map[str]^ToolNode{}}}
  e.currentTool = &e.tools
  e.beginTool("Test")
  e.beginTool("SubTest") // here's a tip, remove this and the issue won't happen
}
```

gives
```
Runtime error /progs/umka-lang/test.um (23): Pointer to a local variable escapes from the function
```

but editor `e` never escapes main as no statics are assigned neither it's returned. May be related to issue #198 as it crashes with sanitizer on

**Comment by skejeton** [comment]
Needless to say if we change the `tools` field to use a pointer allocated with `new` the issue disappears

**Comment by vtereshkov** [comment]
@skejeton You have a circular reference here. To avoid it, please declare `parent` as `^ToolNode`.

**Comment by skejeton** [comment]
Good point, in general garbage collectors are able to handle circular references, but in this case it's a bottleneck of a refcounting GC. This one is a bit of a difficult issue because of that. Right now I have no issue with allocating the `tools` field with `new` or weak pointers as you mentioned, but I think this issue is worth handling as it is a GC language and people using it generally shouldn't worry much about memory. Languages like go, java and c# handle that by virtue of being marking garbage collectors.

**Comment by vtereshkov** [comment]
Blocked by #200.

**Comment by vtereshkov** [comment]
@skejeton I opened a separate issue for a tracing garbage collector. However, it would be a really big change, so I postpone it until Umka 2.x. I have been debugging the current (ref counting) garbage collector for 2.5 years, so I don't expect the tracing garbage collector to require less time.

**Comment by skejeton** [comment]
I think this can be closed.

**Comment by vtereshkov** [comment]
A minimal example that triggers the same error:
```
type T = struct {
        next: ^T
}

var m: ^T

fn main() {
        n := T{}
        n.next = &n
        m = &n   
}                // Runtime error: Pointer to a local variable escapes from the function
```
